network_simulator/link.py

```diff
@@ -95,3 +133,133 @@ def __eq__(self, other):
+        return self.static_delay_sec
+
+    def push(self, packet, dest_dev, global_clock_sec):
+        """
+        Pushes packet onto own LinkBuffer
+
+        :param Packet packet: Packet to push
+        :param Device dest_dev: destination Device of packet
+        :param global_clock_sec: store when packet pushed
+        :return: True if pushed, False if not
+        """
+        return self.link_buffer.push(packet, dest_dev, global_clock_sec)
+
+
+class LinkSendEvent(Event):
+    """
+    Event representing a packet leaving Link's buffer
+    """
+
+    def __init__(self, link):
+        """
+        :ivar link: link that must send its next packet
+        :ivar buff_elem: temporary variable containing LinkBufferElement to send
+        """
+        self.link = link
+        self.buff_elem = None
+
+    def run(self, main_event_loop, statistics):
+        """
+        runs by popping off top packet
+        :param MainEventLoop main_event_loop: event loop for global timing
+        :param Statistics statistics: the Statistics to update
+        """
+
+        # sanity check
+        if self.link.queue.empty():
+            logger.warning("LinkSendEvent while Link empty!")
+            return
+
+        # now pop
+        self.buffer_elem = self.link.pop()
+
+        # update statistics only if not RouterPacket
+        if not isinstance(self.buffer_elem.packet, RouterPacket):
+            statistics.link_packet_transmitted(self.link, \
+                    self.buffer_elem.packet, \
+                    main_event_loop.global_cloc_sec)
+
+    def schedule_new_events(self, main_event_loop):
+        """
+        Schedules new Events. This is called immediately after run().
+        :param MainEventLoop main_event_loop: event loop where new Events will
+        be scheduled.
+        """
+        # schedule a *ReceivedPacketEvent with propagation + transmission
+        # delay, depending on whether Host or Router
+        if isinstance(self.buffer_elem.dest_dev, Router):
+            main_event_loop.schedule_event_with_delay(\
+                    RouterReceivedPacketEvent(0),\
+                    self.link.static_delay_sec + self.packet.size_bits / \
+                        float(self.link.capacity_bps))
+                # TODO(Laksh): RouterReceivedPacketEvent signature
+                # recommended signature RouterReceivedPacketEvent(router, packet)?
+        else:
+            main_event_loop.schedule_event_with_delay(\
+                    HostReceivedPacketEvent(0),\
+                    self.link.static_delay_sec + self.packet.size_bits / \
+                        float(self.link.capacity_bps))
+
+        if self.link.link_buffer.queue.empty():
+            # if link_buffer.queue is empty, then link is no longer busy
+            self.link.busy = False
+        else:
+            # queue new event with just transmission delay
+            assert(self.link.busy == True)
```

Comment:
nit: no need for parentheses around the inside of an assert call. "assert" is not a function in Python.